# **UNLOCKING REAL ESTATE SUCCESS WITH TIME SERIES MODELLING**

This is a collaborative group project done at the end of Phase 4 of Moringa School's Data Science program. The team members of this group include:

- [Abdideq Adan](https://github.com/AdanAbdideq)
- [Clara Gatambia](https://github.com/claragatambia)
- [Isaack Odera](https://github.com/derak-isaack)
- [Mwiti Mwongo](https://github.com/M13Mwongo)
- [Wilson Mutungu](https://github.com/mutungu)

## 1. BUSINESS UNDERSTANDING

TODO - Answer a sample question like why prices would be different in different states but one state may be more expensive on average than another

TODO - Analysis into the market trends (e.g. 2008 financial crash) and how the model may account for that in future (market trend analysis)

TODO - More important to show in depth analysis i.e. data prep and eda

TODO - Think as the investor in this area

### Introduction

### Objectives
The main objective of the project is to establish the 5 best zip codes that are the best to invest in.

The objectives of this project are as follows:
 - Identify the key metrics that would be used to classify profitability.
 - Identify the criteria that would classify a house as a "high-end" house or "affordable" house.

### Potential Challenges

### Conclusion


## PRELIMINARIES

### Loading of Necessary Modules/Packages

All necessary python packages are loaded at once to make the work easy to view and cleaner as well. 

In [1]:
# Importing necessary libraries
# Basics
import pandas as pd
import numpy as np
import itertools
from io import StringIO

# Visualization
import matplotlib.pyplot as plt
import plotly.express as px
import seaborn as sns
import matplotlib.patches as mpatches
from matplotlib.pylab import rcParams
import time

# Modeling
import statsmodels.api as sm
from statsmodels.graphics.tsaplots import plot_acf, plot_pacf
from statsmodels.tsa.statespace.sarimax import SARIMAX
from sklearn.metrics import mean_squared_error
from statsmodels.tsa.seasonal import seasonal_decompose
from statsmodels.tsa.arima.model import ARIMA
from statsmodels.tsa.stattools import acf, pacf, adfuller
from sklearn.linear_model import LassoLarsCV

# Warnings
import warnings
from statsmodels.tools.sm_exceptions import ConvergenceWarning
warnings.simplefilter('ignore', ConvergenceWarning)
warnings.filterwarnings('ignore')

# Custom Options
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns',100)


### Definition of Custom Classes & Functions

In line with the concepts of OOP, custom classes were created to speed up the opening and manipulation of data in this project. They are as follows: 

In [2]:
class DataSourcing:
  def __init__(self, dataframe):
    self.original = dataframe
    self.dataframe = dataframe

  def give_info(self):
    with StringIO() as buffer:
      self.dataframe.info(buf = buffer)
      info_string = buffer.getvalue()
    
    message = f"""
    ----------------------------------------------------------------------->
    DESCRIPTION OF THE DATAFRAME IN QUESTION:
    ----------------------------------------------------------------------->

    Dataframe information => \n{info_string}
    ------------------------------------------------------------------------------------------------------------------------->

    Dataframe shape => {self.dataframe.shape[0]} rows, {self.dataframe.shape[1]} columns
    ------------------------------------------------------------------------------------------------------------------------->

    There are {len(self.dataframe.columns)} columns, namely: {self.dataframe.columns}.
    ------------------------------------------------------------------------------------------------------------------------->

    The first 5 records in the dataframe are seen here:
    ------------------------------------------------------------------------------------------------------------------------->
    {self.dataframe.head()}
    ------------------------------------------------------------------------------------------------------------------------->

    The last 5 records in the self.dataframe are as follows:
    ------------------------------------------------------------------------------------------------------------------------->
    {self.dataframe.tail()}
    ------------------------------------------------------------------------------------------------------------------------->

    The descriptive statistics of the dataframe (mean,median, max, min, std) are as follows:
    ------------------------------------------------------------------------------------------------------------------------->
    {self.dataframe.describe()}
    ------------------------------------------------------------------------------------------------------------------------->
    """
    print(message)

  def null_count(self):
    return print(self.dataframe.isnull().sum())

  def unique_count(self):
    return self.dataframe.nunique()

  def unique_per_column(self):
    print("<----- UNIQUE VALUES IN EACH COLUMN ----->")
    for col in self.dataframe.columns:
      print(f"{col} : \n {self.dataframe[col].unique()}")
      print()
    print("<----- END OF UNIQUE VALUES IN EACH COLUMN ----->")
    return

  def plot_barplot(self, dataframe, x, y, x_title, y_title):
    fig, ax = plt.subplots(figsize=(10, 8))

    sns.barplot(data=dataframe, x=x, y=y, orient='h', errorbar=None)
    ax.set_title(f"{y_title} vs {x_title}")
    ax.set_xlabel(x_title)
    ax.set_ylabel(y_title)
    plt.show()

  def plot_boxplot(self, dataframe, x, y, x_title, y_title):
    fig, ax = plt.subplots(figsize=(10, 8))

    sns.boxplot(data=dataframe, x=x, y=y)
    ax.set_title(f"{y_title} vs {x_title}")
    ax.set_xlabel(x_title)
    ax.set_ylabel(y_title)
    plt.xticks(rotation=90)
    plt.show()

In [3]:
class DataPreProcessing(DataSourcing):
  def __init__(self, dataframe):
    super().__init__(dataframe)

  def dropColumns(self, columns):
    return self.dataframe.drop(columns, axis=1)

  def dropRows(self, rows):
    return self.dataframe.drop(rows, axis=0)

## 2. DATA UNDERSTANDING

### a) Loading & Viewing Data

We start by loading the data as a dataframe.

In [4]:
df = pd.read_csv('./data/zillow_data.csv')

The `df.head()` function is used to get a rough look at a few of the records in the dataframe to understand the data better.

In [5]:
df.head(10)

RegionID  RegionName           City State              Metro  \
0     84654       60657        Chicago    IL            Chicago   
1     90668       75070       McKinney    TX  Dallas-Fort Worth   
2     91982       77494           Katy    TX            Houston   
3     84616       60614        Chicago    IL            Chicago   
4     93144       79936        El Paso    TX            El Paso   
5     91733       77084        Houston    TX            Houston   
6     61807       10467       New York    NY           New York   
7     84640       60640        Chicago    IL            Chicago   
8     91940       77449           Katy    TX            Houston   
9     97564       94109  San Francisco    CA      San Francisco   

      CountyName  SizeRank   1996-04   1996-05   1996-06   1996-07   1996-08  \
0           Cook         1  334200.0  335400.0  336500.0  337600.0  338500.0   
1         Collin         2  235700.0  236900.0  236700.0  235400.0  233300.0   
2         Harris         3  210400.0  212200.0  212200.0  210700.0  208300.0   
3           Cook         4  498100.0  500900.0  503100.0  504600.0  505500.0   
4        El Paso         5   77300.0   77300.0   77300.0   77300.0   77400.0   
5         Harris         6   95000.0   95200.0   95400.0   95700.0   95900.0   
6          Bronx         7  152900.0  152700.0  152600.0  152400.0  152300.0   
7           Cook         8  216500.0  216700.0  216900.0  217000.0  217100.0   
8         Harris         9   95400.0   95600.0   95800.0   96100.0   96400.0   
9  San Francisco        10  766000.0  771100.0  776500.0  781900.0  787300.0   

    1996-09   1996-10   1996-11   1996-12   1997-01   1997-02   1997-03  \
0  339500.0  340400.0  341300.0  342600.0  344400.0  345700.0  346700.0   
1  230600.0  227300.0  223400.0  219600.0  215800.0  211100.0  205700.0   
2  205500.0  202500.0  199800.0  198300.0  197300.0  195400.0  193000.0   
3  505700.0  505300.0  504200.0  503600.0  503400.0  502200.0  500000.0   
4   77500.0   77600.0   77700.0   77700.0   77800.0   77900.0   77900.0   
5   96100.0   96200.0   96100.0   96000.0   95800.0   95500.0   95300.0   
6  152000.0  151800.0  151600.0  151600.0  151700.0  151800.0  151800.0   
7  217200.0  217500.0  217900.0  218600.0  219700.0  220900.0  221800.0   
8   96700.0   96800.0   96800.0   96700.0   96600.0   96400.0   96200.0   
9  793000.0  799100.0  805800.0  814400.0  824300.0  833800.0  842900.0   

    1997-04   1997-05   1997-06   1997-07   1997-08   1997-09   1997-10  \
0  347800.0  349000.0  350400.0  352000.0  353900.0  356200.0  358800.0   
1  200900.0  196800.0  193600.0  191400.0  190400.0  190800.0  192700.0   
2  191800.0  191800.0  193000.0  195200.0  198400.0  202800.0  208000.0   
3  497900.0  496300.0  495200.0  494700.0  494900.0  496200.0  498600.0   
4   77800.0   77800.0   77800.0   77800.0   77800.0   77900.0   78100.0   
5   95100.0   95100.0   95200.0   95400.0   95600.0   95800.0   96000.0   
6  151900.0  152000.0  152200.0  152400.0  152500.0  152600.0  152700.0   
7  223000.0  224200.0  225600.0  227100.0  228800.0  230600.0  232700.0   
8   96100.0   96200.0   96300.0   96600.0   97000.0   97500.0   98000.0   
9  852900.0  863500.0  874800.0  886500.0  898200.0  910200.0  922800.0   

    1997-11   1997-12   1998-01   1998-02    1998-03    1998-04    1998-05  \
0  361800.0  365700.0  370200.0  374700.0   378900.0   383500.0   388300.0   
1  196000.0  201300.0  207400.0  212200.0   214600.0   215100.0   213400.0   
2  213800.0  220700.0  227500.0  231800.0   233400.0   233900.0   233500.0   
3  502000.0  507600.0  514900.0  522200.0   529500.0   537900.0   546900.0   
4   78200.0   78400.0   78600.0   78800.0    79000.0    79100.0    79200.0   
5   96200.0   96400.0   96500.0   96600.0    96600.0    96700.0    96900.0   
6  152900.0  153200.0  153800.0  154300.0   154700.0   155200.0   155700.0   
7  235000.0  237900.0  241300.0  244600.0   247900.0   251300.0   254800.0   
8   98400.0   98800.0   99200.

The `df.info()` and `df.dtypes` functions are both called to give a rough understanding of the dataframe, and the types of data held in each column. Normally, `df.info()` would be sufficient, but due to the sheer number of columns in the dataframe, that information isn't displayed. Thus, `df.dtypes` is called. 

In [6]:
print(df.info())
print()
print(df.dtypes)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14723 entries, 0 to 14722
Columns: 272 entries, RegionID to 2018-04
dtypes: float64(219), int64(49), object(4)
memory usage: 30.6+ MB
None

RegionID        int64
RegionName      int64
City           object
State          object
Metro          object
CountyName     object
SizeRank        int64
1996-04       float64
1996-05       float64
1996-06       float64
1996-07       float64
1996-08       float64
1996-09       float64
1996-10       float64
1996-11       float64
1996-12       float64
1997-01       float64
1997-02       float64
1997-03       float64
1997-04       float64
1997-05       float64
1997-06       float64
1997-07       float64
1997-08       float64
1997-09       float64
1997-10       float64
1997-11       float64
1997-12       float64
1998-01       float64
1998-02       float64
1998-03       float64
1998-04       float64
1998-05       float64
1998-06       float64
1998-07       float64
1998-08       float64
1998-09       floa

As expected, all the columns representing time-series data will be storing either an integer or a float data type. It is also important to note that the `RegionID` & `RegionName` are stored as integers - to ensure they are unique - and the remaining 5 columns all contain strings.

The shape of the dataframe was observed as follows: 

In [7]:
df.shape

(14723, 272)

The dataframe has 14723 records and 272 column. However, majority of those columns represent the time series values. The first seven columns - `RegionID`, `RegionName`, `City`, `State`, `Metro`, `CountyName` and `SizeRank` - are the columns that give us more information about the dataset. 

Each of these columns is intended to hold a certain type of data as follows: 
- *RegionID*: The unique ID of the region in question.
- *RegionName*: The name of the region in question.
- *City*: The name of the city within a given region.
- *State*: The state in which the RegionID is found.
- *Metro*: The metropolitan name within which the RegionID is found.
- *CountyName*: The name of the county within a given region.
- *SizeRank*: The region's area ranking vis-a-vis other regions, organised in descending order.

Going forward, these columns shall be referred to as the columns of interest.

In [8]:
# ['RegionID', 'RegionName', 'City', 'State', 'Metro', 'CountyName','SizeRank']

cols_of_interest = df.columns[:7]
cols_of_interest

Index(['RegionID', 'RegionName', 'City', 'State', 'Metro', 'CountyName',
       'SizeRank'],
      dtype='object')

### b) Data Cleaning

#### **i. Handling of Null Values**

The number of null values in the dataframe will now be established. This is necessary to ensure an appropriate strategy is used when deciding how to handle null values

In [9]:
df.isnull().sum()

RegionID         0
RegionName       0
City             0
State            0
Metro         1043
CountyName       0
SizeRank         0
1996-04       1039
1996-05       1039
1996-06       1039
1996-07       1039
1996-08       1039
1996-09       1039
1996-10       1039
1996-11       1039
1996-12       1039
1997-01       1039
1997-02       1039
1997-03       1039
1997-04       1039
1997-05       1039
1997-06       1039
1997-07       1038
1997-08       1038
1997-09       1038
1997-10       1038
1997-11       1038
1997-12       1038
1998-01       1036
1998-02       1036
1998-03       1036
1998-04       1036
1998-05       1036
1998-06       1036
1998-07       1036
1998-08       1036
1998-09       1036
1998-10       1036
1998-11       1036
1998-12       1036
1999-01       1036
1999-02       1036
1999-03       1036
1999-04       1036
1999-05       1036
1999-06       1036
1999-07       1036
1999-08       1036
1999-09       1036
1999-10       1036
1999-11       1036
1999-12       1036
2000-01     

Aside from the columns that contain time-series values, only the metro column has null values. This can be seen as inconsequential as there is enough data from the other columns to overlook this. 

Taking a closer look at the missing time-series values, there seems to be a steady trend. The values do not appear to be random, as they steadily decrease from April 1996 to July 2014, from where they remain 0 throughout till April 2018.

The gradual change indicates that the presence of null values in these columns is anything but random. As such, when dealing with the null values in the time-series value columns, **all null values will be left as is**. No replacement or removal of records will be done. This is done as it is assumped that not all the houses were built at the same time, thus it is expected that there will be null values for some houses and not others. Furthermore, some of these null values are attributed to the differential times that the houses were put on the market.

When dealing with the null values in the metro column, all null values will be left as the data missing does not impact the objectives of the project. 



#### **ii. Duplicate Values Check**

A check for duplicate values is done. This check specifically applies to the non time-series columns, as it possible to have duplicates in the value of a property, and such duplicates do not serve to understand the dataset better.

In [10]:
duplicates = df.duplicated(subset=cols_of_interest,keep=False)

df[duplicates]

Empty DataFrame
Columns: [RegionID, RegionName, City, State, Metro, CountyName, SizeRank, 1996-04, 1996-05, 1996-06, 1996-07, 1996-08, 1996-09, 1996-10, 1996-11, 1996-12, 1997-01, 1997-02, 1997-03, 1997-04, 1997-05, 1997-06, 1997-07, 1997-08, 1997-09, 1997-10, 1997-11, 1997-12, 1998-01, 1998-02, 1998-03, 1998-04, 1998-05, 1998-06, 1998-07, 1998-08, 1998-09, 1998-10, 1998-11, 1998-12, 1999-01, 1999-02, 1999-03, 1999-04, 1999-05, 1999-06, 1999-07, 1999-08, 1999-09, 1999-10, 1999-11, 1999-12, 2000-01, 2000-02, 2000-03, 2000-04, 2000-05, 2000-06, 2000-07, 2000-08, 2000-09, 2000-10, 2000-11, 2000-12, 2001-01, 2001-02, 2001-03, 2001-04, 2001-05, 2001-06, 2001-07, 2001-08, 2001-09, 2001-10, 2001-11, 2001-12, 2002-01, 2002-02, 2002-03, 2002-04, 2002-05, 2002-06, 2002-07, 2002-08, 2002-09, 2002-10, 2002-11, 2002-12, 2003-01, 2003-02, 2003-03, 2003-04, 2003-05, 2003-06, 2003-07, 2003-08, 2003-09, 2003-10, 2003-11, 2003-12, ...]
Index: []

[0 rows x 272 columns]

No duplicates are observed in the dataset.

#### **iii. Consistent Data Types**

Having run `df.dtypes` earlier, it was noted that some of the price data in the time-series columns was saved as integers while others were saved as floats. All the data in the time-series columns will now be converted to floats since by doing so, we are sure we are not losing/ rounding off any of the price values. 

In [11]:
# Ensuring all time series columns are converted to float64 i.e. the last 265 columns

for col in df.columns[-265:]:
  if df[col].dtype is not 'float64':
    df[col] = df[col].astype('float64')


df.dtypes

RegionID        int64
RegionName      int64
City           object
State          object
Metro          object
CountyName     object
SizeRank        int64
1996-04       float64
1996-05       float64
1996-06       float64
1996-07       float64
1996-08       float64
1996-09       float64
1996-10       float64
1996-11       float64
1996-12       float64
1997-01       float64
1997-02       float64
1997-03       float64
1997-04       float64
1997-05       float64
1997-06       float64
1997-07       float64
1997-08       float64
1997-09       float64
1997-10       float64
1997-11       float64
1997-12       float64
1998-01       float64
1998-02       float64
1998-03       float64
1998-04       float64
1998-05       float64
1998-06       float64
1998-07       float64
1998-08       float64
1998-09       float64
1998-10       float64
1998-11       float64
1998-12       float64
1999-01       float64
1999-02       float64
1999-03       float64
1999-04       float64
1999-05       float64
1999-06   

### c) Data Handling

#### **i. Aggregation of Data**

As all values in the time-series columns are deemed important, aggregating the data to a yearly basis, for example, at such an early stage was deemed unnecessary. This is because a lot of vital information would be lost in the process that would be necessary later on. Thus, aggregation would be done on an as-needed basis.

Data will now be grouped by state, to give a better idea as to which states have more property listings than others:

### d) Feature Engineering

#### **i. Addition of New Features** 

Relevant features that could affect house prices were noted and their creation was deemed necessary. These features are: 

- *Return On Investment(%)*: Calculated as 
$
\left( \frac{{\text{{Last Price of Property}}}}{{\text{{Initial Price of Property}}}} - 1 \right) \times 100\% 
$


-

In [ ]:
df[['State','RegionID','RegionName','City','Metro','CountyName','SizeRank']].groupby("State").count()

Particular focus will be placed on the `RegionID` as that column does not have any null values, thus an accurate count of properties can be done.

In [ ]:
df_state_grouping = df[['State','RegionID']].groupby("State").count().sort_values(by='RegionID', ascending=False)
df_state_grouping

California (CA) is seen to have the most listings with 1224, followed closely by New York, NY (1015) and Texas, TX (989). Vermont (VT), Washington DC (DC) and San Diego (SD) have the least listings at 16, 18 and 19 respectively. 

A new dataframe is now created which contains the melted data. This will convert the original dataframe `df` from wide to long format. This will be done by applying the custom function `melt_data`

In [ ]:
def melt_data(df):
    """
    Takes the zillow_data dataset in wide form or a subset of the zillow_dataset.  
    Returns a long-form datetime dataframe 
    with the datetime column names as the index and the values as the 'values' column.
    
    If more than one row is passes in the wide-form dataset, the values column
    will be the mean of the values from the datetime columns in all of the rows.
    """

    melted = pd.melt(df, id_vars=['RegionName', 'RegionID', 'SizeRank',
                     'City', 'State', 'Metro', 'CountyName'], var_name='time')
    melted['time'] = pd.to_datetime(melted['time'], infer_datetime_format=True)
    melted = melted.dropna(subset=['value'])
    return melted.groupby('time').aggregate({'value': 'mean'})

Converting the dataframe to long format:

In [ ]:
df_melted = melt_data(df)

## 3. EDA & VISUALISATION

### Descriptive Statistics

## 4. FEATURE ENGINEERING

### Dealing with Missing Data

### Dealing with Outliers

### Scaling & Normalization

### Extracting Time & Date Features

## 5. ARIMA MODELLING

## 6. FINE TUNING WITH PROPHET

## 7. INTERPRETING RESULTS